karate-core/src/main/java/com/intuit/karate/StepDefs.java

```diff
@@ -43,309 +33,327 @@
+		handleFailure(ar);
+	}
+
+	@When("^method (\\w+)")
+	public void method(String method) {
+		request.setMethod(method);
+		response = context.client.invoke(request, context);
+		context.vars.put(ScriptValueMap.VAR_RESPONSE_STATUS, response.getStatus());
+		context.vars.put(ScriptValueMap.VAR_RESPONSE_TIME, response.getTime());
+		context.vars.put(ScriptValueMap.VAR_COOKIES, response.getCookies());
+		if (response.getHeaders() != null) {
+			DocumentContext headers = JsonPath.parse(response.getHeaders());
+			context.vars.put(ScriptValueMap.VAR_RESPONSE_HEADERS, headers);
+		} else {
+			context.vars.put(ScriptValueMap.VAR_RESPONSE_HEADERS, ScriptValue.NULL);
+		}
+		Object responseBody = convertResponseBody(response.getBody());
+		if (responseBody instanceof String) {
+			String responseString = (String) responseBody;
+			if (Script.isJson(responseString)) {
+				responseBody = JsonUtils.toJsonDoc(responseString);
+			} else if (Script.isXml(responseString)) {
+				try {
+					responseBody = XmlUtils.toXmlDoc(responseString);
+				} catch (Exception e) {
+					logger.warn("xml parsing failed, response data type set to string: {}", e.getMessage());
+				}
+			}
+		}
+		context.vars.put(ScriptValueMap.VAR_RESPONSE, responseBody);
+		String prevUrl = request.getUrl();
+		Map<String, Cookie> prevCookies = request.getCookies();
+		request = new HttpRequest();
+		request.setUrl(prevUrl);
+		if (prevCookies == null) {
+			request.setCookies(response.getCookies());
+		} else {
+			if (response.getCookies() != null) {
+				prevCookies.putAll(response.getCookies());
+			}
+			request.setCookies(prevCookies);
+		}
+	}
+
+	private static Object convertResponseBody(byte[] bytes) {
+		if (bytes == null) {
+			return null;
+		}
+		// if a byte array contains a negative-signed byte,
+		// then the string conversion will corrupt it.
+		// in that case just return the byte array stream
+		try {
+			String rawString = new String(bytes, "utf-8");
+			if (Arrays.equals(bytes, rawString.getBytes())) {
+				return rawString;
+			}
+		} catch (Exception e) {
+			logger.warn("response bytes to string conversion failed: {}", e.getMessage());
+		}
+		return new ByteArrayInputStream(bytes);
+	}
+
+	@When("^soap action( .+)?")
+	public void soapAction(String action) {
+		action = Script.eval(action, context).getAsString();
+		if (action == null) {
+			action = "";
+		}
+		request.addHeader("SOAPAction", action);
+		request.addHeader("Content-Type", "text/xml");
+		method("post");
+	}
+
+	@When("^multipart entity (.+)")
+	public void multiPartEntity(String value) {
+		multiPart(null, value);
+	}
+
+	@When("^multipart field (.+) = (.+)")
+	public void multiPartFormField(String name, String value) {
+		multiPart(name, value);
+	}
+
+	public void multiPart(String name, String value) {
+		ScriptValue sv = Script.eval(value, context);
+		request.addMultiPartItem(name, sv);
+	}
+
+	@Then("^print (.+)")
+	public void print(List<String> expList) {
```

Comment:
probably better to use stringbuilder
would be good to have a few examples (e.g. in syntax.feature) to demo / unit-test this